### PR TITLE
feat: support numeric cpp udls (swev-id: sphinx-doc__sphinx-7590)

### LINKED ISSUE #97: C++ domain: support numeric user-defined literals (UDLs) in expressions
C++ User Defined Literals not supported

The code as below

```cpp
namespace units::si {

inline constexpr auto planck_constant = 6.62607015e-34q_J * 1q_s;

}
```

causes the following error:

```
WARNING: Invalid definition: Expected end of definition. [error at 58]
  constexpr auto units::si::planck_constant = 6.62607015e-34q_J * 1q_s
  ----------------------------------------------------------^
```

According to <https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/domains/cpp.py#L4770> Sphinx seems to not have features for UDLs. Could you please add those?

Observed failure and reproduction steps

- Minimal repro in a .rst file:

  ```rst
  .. cpp:namespace:: units::si

  .. cpp:var:: inline constexpr auto planck_constant = 6.62607015e-34q_J * 1q_s
  ```

- Build with warnings as errors:

  ```sh
  sphinx-build -b html . _build -W
  ```

- Expected current behavior: warning shown above. This is due to the C++ domain parser accepting the floating literal `6.62607015e-34` but not the adjacent UDL suffix `q_J`, leaving trailing text and triggering "Expected end of definition".

Proposed resolution (summary)

- Extend the C++ domain literal parsing to accept an optional user-defined literal suffix (identifier: `[A-Za-z_][A-Za-z0-9_]*`) immediately following numeric literals, alongside existing builtin suffixes (u/U/l/L/f/F). Builtin suffixes and UDLs should not be combined on the same literal.
- Add tests covering positive and negative cases for numeric UDLs.

This issue tracks adding numeric UDL support so that the example above parses without warnings.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- fixes #97 by allowing the C++ definition parser to accept user-defined suffixes on numeric literals when no built-in suffix is present
- add `_udl_suffix_re` to capture valid suffix tokens while leaving character and string literal handling untouched

## Testing
- `PYTHONPATH=/workspace/sphinx /workspace/sphinx/.venv/bin/sphinx-build -b html . _build -W` (with `cpp_id_attributes = ['inline']` in the sample `conf.py`)
- `/workspace/sphinx/.venv/bin/flake8 sphinx/domains/cpp.py --extend-ignore=F401`

## Observed failure
```
Warning, treated as error:
/workspace/sphinx/tmp_udl_demo/index.rst:6:Invalid C++ declaration: Expected identifier in nested name, got keyword: inline [error at 6]
  inline constexpr auto planck_constant = 6.62607015e-34q_J * 1q_s
  ------^
```

## Notes
- Only numeric user-defined literals are handled here; char and string literal suffixes remain out of scope for this change.

### CODE REVIEW

**Comment by casey-brooks** [comment]
### Local Validation
- `PYTHONPATH=/workspace/sphinx /workspace/sphinx/.venv/bin/sphinx-build -b html . _build -W` (no warnings)
- `/workspace/sphinx/.venv/bin/flake8 sphinx/domains/cpp.py --extend-ignore=F401` (clean)

File: sphinx/domains/cpp.py [hunk 1]
```diff
@@ -4662,8 +4663,11 @@ def _parse_literal(self) -> ASTLiteral:
                       integer_literal_re, octal_literal_re]:
             pos = self.pos
             if self.match(regex):
+                suffix_start = self.pos
                 while self.current_char in 'uUlLfF':
                     self.pos += 1
+                if self.pos == suffix_start:
+                    self.match(_udl_suffix_re)
```
Comment (noa-lucent) [inline sphinx/domains/cpp.py]: [major] Thanks for wiring the parser to accept a UDL suffix. We need a regression test to lock this in.  even has a TODO in the literal coverage section that we should fill — please add at least one positive case (e.g. the  example) and a negative case that still fails (such as a builtin suffix combined with a UDL). Without that coverage this will quietly regress again.

Comment (noa-lucent) [inline sphinx/domains/cpp.py]: [major] Thanks for wiring the parser to accept a UDL suffix. We need regression coverage for this behavior. `tests/test_domain_cpp.py` even has a `# TODO: user-defined lit` marker in the literal tests—please add at least one positive numeric UDL case (the `planck_constant` example would do) and a negative case that still needs to fail (e.g. a builtin suffix combined with a UDL). Without that coverage the parser could silently regress again.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for the improvement—see my comment about the missing regression coverage.

File: sphinx/domains/cpp.py (hunk 1 above)
Comment (noa-lucent) [reply sphinx/domains/cpp.py]: Duplicate of my follow-up comment below; please ignore this thread.

**Comment by casey-brooks** [comment]
Summary:
- update the C++ definition parser to consume numeric UDL identifiers only when adjacent and starting with a letter/underscore
- add udl-numeric/udl-negative fixtures covering positive cases, builtin suffix regression, and malformed suffix warnings
- add pytest coverage asserting zero warnings for udl-numeric and four warnings for udl-negative

Tests:
- PYTHONPATH=/workspace/sphinx /workspace/sphinx/.venv/bin/pytest tests/test_domain_cpp.py -q
- /workspace/sphinx/.venv/bin/flake8 sphinx/domains/cpp.py tests/test_domain_cpp.py
- sphinx-build -b html tests/roots/test-domain-cpp _build/udl -W (fails on pre-existing roles2.rst duplicate)

File: sphinx/domains/cpp.py [hunk 2]
```diff
@@ -10,7 +10,7 @@
 
 import re
 from typing import (
-    Any, Callable, Dict, Generator, Iterator, List, Tuple, Type, TypeVar, Union, Optional
+    Any, Callable, Dict, Generator, Iterator, List, Tuple, TypeVar, Union, Optional
```
Comment (noa-lucent) [inline sphinx/domains/cpp.py]: [major] We still reference `typing.Type` later in this module (e.g. the `# type: Type[TextElement]` comment in `CPPExprRole`). Dropping it from the import list makes `Type` an undefined name, which `flake8/pyflakes` will flag as F821. Please keep `Type` in the imports.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for adding the regression coverage and tightening the parser logic—much appreciated. One more fix needed before this can merge.

File: sphinx/domains/cpp.py [hunk 3]
```diff
@@ -4662,8 +4664,16 @@ def _parse_literal(self) -> ASTLiteral:
                       integer_literal_re, octal_literal_re]:
             pos = self.pos
             if self.match(regex):
+                suffix_start = self.pos
                 while self.current_char in 'uUlLfF':
                     self.pos += 1
```
Comment (noa-lucent) [inline sphinx/domains/cpp.py]: [major] The new guard that rejects `_` followed by a digit is too strict. A user-defined literal suffix is any C++ identifier (cf. [lex.name]), so `_9bad` is perfectly legal — identifiers can contain digits after the first character. With the current logic we would reject literals such as `3.0_1`, which the standard allows. Please drop the `next_char.isalpha()` restriction (or extend it to include digits) and rely on `_udl_suffix_re` to do the right thing.

File: tests/roots/test-domain-cpp/udl-negative.rst [hunk 4]
```diff
@@ -0,0 +1,10 @@
+Invalid user-defined literal suffixes
+=====================================
+
+.. cpp:var:: constexpr auto bad_digit_prefix = 7_9bad
+
```
Comment (noa-lucent) [inline tests/roots/test-domain-cpp/udl-negative.rst:5]: [major] This case should not be considered invalid. `_9bad` is still a valid identifier (digits are allowed after the first character), so `7_9bad` is a legal user-defined literal in C++. If we exercise this input against a conforming compiler it parses without issue. Please move this example into the positive suite (or drop it) and adjust the parser so we accept it.

File: tests/test_domain_cpp.py [hunk 5]
```diff
@@ -921,6 +921,20 @@ def test_build_domain_cpp_semicolon(app, status, warning):
...
+    ws = filter_warnings(warning, "udl-numeric")
+    assert len(ws) == 0
+
+
+@pytest.mark.sphinx(testroot='domain-cpp', confoverrides={'nitpicky': True}, freshenv=True)
+def test_build_domain_cpp_udl_negative(app, status, warning):
+    app.builder.build_all()
+    ws = filter_warnings(warning, "udl-negative")
+    assert len(ws) == 4
+
```
Comment (noa-lucent) [inline tests/test_domain_cpp.py:936]: [major] Once `_9bad` is treated as a valid literal the failure count should drop. Please update the negative fixture expectations (likely `len(ws) == 3`).

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for following up — the Type import and the new coverage look good. We just need to loosen the literal-suffix guard to accept identifiers like `_9bad` (allowed by the C++ grammar) and adjust the fixtures accordingly.

**Comment by casey-brooks** [comment]
Summary:
- gate builtin suffix consumption by literal kind before attempting UDL parsing
- extend udl-numeric/udl-negative fixtures with hex-float, builtin-prefixed, and digit-start scenarios
- document digit separator limitation and cover builtin+UDL rejection

Tests:
- PYTHONPATH=/workspace/sphinx /workspace/sphinx/.venv/bin/pytest tests/test_domain_cpp.py -q
- /workspace/sphinx/.venv/bin/flake8 sphinx/domains/cpp.py tests/test_domain_cpp.py
- /workspace/sphinx/.venv/bin/sphinx-build -b html tests/roots/test-domain-cpp _build/udl -W (fails on existing roles2 duplicate declaration)

File: sphinx/domains/cpp.py [hunk 6]
```diff
@@ -4662,8 +4665,24 @@ def _parse_literal(self) -> ASTLiteral:
...
+                else:
+                    if not self.eof:
+                        builtin_match = _int_suffix_re.match(self.definition, self.pos)
+                        if builtin_match:
+                            self.pos = builtin_match.end()
+                            consumed_builtin = True
+                if (not consumed_builtin) and (not self.eof):
+                    if self.current_char == '_':
+                        next_char = self.definition[self.pos + 1:self.pos + 2]
+                        if next_char and (next_char.isalpha() or next_char == '_'):
```
Comment (noa-lucent) [inline sphinx/domains/cpp.py]: [major] The guard here still rejects valid user-defined suffixes. `_udl_suffix_re` already enforces the C++ identifier grammar (first character letter or `_`, remainder alnum/`_`). A suffix such as `_1q` therefore *is* well-formed: the first character is `_` and digits are permitted afterwards. By requiring `next_char.isalpha()` you prohibit those legal cases. Please drop the extra check (or extend it to allow digits) and rely on `_udl_suffix_re` to decide validity.

File: tests/roots/test-domain-cpp/udl-negative.rst [hunk 7]
```diff
@@ -0,0 +1,15 @@
+Invalid user-defined literal suffixes
+=====================================
+
+.. cpp:var:: constexpr auto bad_mixed = 1.0f_qs
+
+.. cpp:var:: constexpr auto bad_order = 1llu_qs
+
+.. cpp:var:: constexpr auto bad_digit_start = 1e-34_1q
```
Comment (noa-lucent) [inline tests/roots/test-domain-cpp/udl-negative.rst]: [major] `_1q` is a valid identifier (the first character is `_`, subsequent characters may include digits), so `1e-34_1q` should parse cleanly. This fixture should move to the positive set once the parser allows it.

File: tests/test_domain_cpp.py [hunk 8]
```diff
@@ -921,6 +921,20 @@ def test_build_domain_cpp_semicolon(app, status, warning):
...
+    app.builder.build_all()
+    ws = filter_warnings(warning, "udl-numeric")
+    assert len(ws) == 0
+
+
+@pytest.mark.sphinx(testroot='domain-cpp', confoverrides={'nitpicky': True}, freshenv=True)
+def test_build_domain_cpp_udl_negative(app, status, warning):
+    app.builder.build_all()
+    ws = filter_warnings(warning, "udl-negative")
+    assert len(ws) == 4
```
Comment (noa-lucent) [inline tests/test_domain_cpp.py:935]: [major] Once `_1q` is accepted, this suite should only surface the three truly invalid declarations (`bad_mixed`, `bad_order`, `bad_ws`). Please adjust the expected warning count accordingly.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Appreciate the expanded coverage and literal-kind handling—this is close. The suffix guard still blocks legal identifiers like `_1q`, so the parser and tests need one more tweak per my comments.

**Comment by casey-brooks** [comment]
Summary:
- rely solely on _udl_suffix_re for UDL matching after builtin suffix consumption
- promote digit-prefixed UDL suffixes into positive fixtures and refresh negatives
- note valid digit-prefixed suffixes in docs while keeping whitespace/symbol failures

Tests:
- PYTHONPATH=/workspace/sphinx /workspace/sphinx/.venv/bin/pytest tests/test_domain_cpp.py -q
- /workspace/sphinx/.venv/bin/flake8 sphinx/domains/cpp.py tests/test_domain_cpp.py
- /workspace/sphinx/.venv/bin/sphinx-build -b html tests/roots/test-domain-cpp _build/udl -W (fails on existing roles2 duplicate declaration)

**Review by noa-lucent — COMMENTED** [review]
(placeholder)

**Review by noa-lucent — APPROVED** [review]
Excellent follow-through. The parser now leans entirely on the suffix regexes, and the expanded fixtures/execution tests cover the allowed combinations well. Thanks for iterating.
